File: 63. Unique Paths II/uniquePathsWithObstacles.js
/**
 * @param {number[][]} obstacleGrid
 * @return {number}
 */
const uniquePathsWithObstacles = (obstacleGrid) => {
  const M = obstacleGrid.length;
  const N = obstacleGrid[0].length;

  if (obstacleGrid[0][0] === 1) {
    return 0;
  }

  obstacleGrid[0][0] = 1;

  for (let i = 1; i < M; i++) {
    obstacleGrid[i][0] = obstacleGrid[i][0] === 1 ? 0 :  obstacleGrid[i - 1][0];
  }

  for (let j = 1; j < N; j++) {
    obstacleGrid[0][j] = obstacleGrid[0][j] === 1 ? 0 :  obstacleGrid[0][j - 1];
  }

  for (let i = 1; i < M; i++) {
    for (let j = 1; j < N; j++) {
      if (obstacleGrid[i][j] === 1) {
        obstacleGrid[i][j] = 0;
      } else {
        obstacleGrid[i][j] = obstacleGrid[i - 1][j] + obstacleGrid[i][j - 1];
      }
    }
  }
  
  return obstacleGrid[M - 1][N - 1];
};

const obstacleGrid = [
  [0, 0, 0],
  [0, 1, 0],
  [0, 0, 0],
];
console.log("uniquePathsWithObstacles: ", uniquePathsWithObstacles(obstacleGrid)); // 2

File: 63. Unique Paths II/uniquePathsWithObstacles.py
from typing import List

class Solution:
  def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
    M, N = len(obstacleGrid), len(obstacleGrid[0])
    dp = [0] * N
    dp[0] = 1

    for i in range(M):
      for j in range(N):
        if obstacleGrid[i][j] == 1:
          dp[j] = 0
        elif j > 0:
          dp[j] += dp[j - 1]

    return dp[-1]
  

obstacleGrid = [
  [0, 0, 0],
  [0, 1, 0],
  [0, 0, 0]
]
solution = Solution()
print('uniquePathsWithObstacles: ', solution.uniquePathsWithObstacles(obstacleGrid)) # 2
